lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -168,51 +180,70 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
 
 
 def path_to_3d_segment(path, zs=0, zdir='z'):
-    """Convert a path to a 3D segment."""
+    """Convert a path to a 3D segment.
 
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
+    """
     zs = np.broadcast_to(zs, len(path))
+
     pathsegs = path.iter_segments(simplify=False, curves=False)
-    seg = [(x, y, z) for (((x, y), code), z) in zip(pathsegs, zs)]
-    seg3d = [juggle_axes(x, y, z, zdir) for (x, y, z) in seg]
-    return seg3d
+    if len(path):
+        xs, ys = zip(*((x, y) for (x, y), code in pathsegs))
+    else:
+        xs, ys = [], []
+
+    seg3d = juggle_axes(xs, ys, zs, zdir)
+    return np.array(seg3d).T
 
 
 def paths_to_3d_segments(paths, zs=0, zdir='z'):
-    """Convert paths from a collection object to 3D segments."""
+    """Convert paths from a collection object to 3D segments.
+
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
+    """
 
     zs = np.broadcast_to(zs, len(paths))
+
     segs = [path_to_3d_segment(path, pathz, zdir)
             for path, pathz in zip(paths, zs)]
     return segs
 
 
 def path_to_3d_segment_with_codes(path, zs=0, zdir='z'):
-    """Convert a path to a 3D segment with path codes."""
+    """Convert a path to a 3D segment with path codes.
 
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
+    """
     zs = np.broadcast_to(zs, len(path))
-    seg = []
-    codes = []
+
     pathsegs = path.iter_segments(simplify=False, curves=False)
-    for (((x, y), code), z) in zip(pathsegs, zs):
-        seg.append((x, y, z))
-        codes.append(code)
-    seg3d = [juggle_axes(x, y, z, zdir) for (x, y, z) in seg]
-    return seg3d, codes
+    codes = np.empty(len(path))
+    xs, ys, codes = zip(*((x, y, code) for (x, y), code in pathsegs))
+    seg3d = juggle_axes(xs, ys, zs, zdir)
+    return np.array(seg3d).T, codes
```

Comment:
Similarly, this would become just a `.view` by a dtype with the fields at different offsets, possibly eliminating a copy